kernel/sched.c

```diff
@@ -1050,6 +1050,25 @@ static inline void z_vrfy_k_thread_deadline_set(k_tid_t tid, int deadline)
 #endif /* CONFIG_USERSPACE */
 #endif /* CONFIG_SCHED_DEADLINE */
 
+void z_impl_k_reschedule(void)
+{
+	k_spinlock_key_t key;
+
+	key = k_spin_lock(&_sched_spinlock);
+
+	update_cache(0);
+	flag_ipi(IPI_ALL_CPUS_MASK);
```

Comment:
Broadcasting an IPI here seems very heavyweight.  The use case at hand is just getting the scheduler to do a reschedule on ISR exit, nothing to do with SMP behavior.  Basically this implements "do an immediate reschedule on all cores in the system", which is a lot more work than required.